File: apps/web/lib/crypto.ts
import nacl from "tweetnacl";
import { canonicalJson } from "@people/shared";

/** Convert Uint8Array to hex string */
export function toHex(bytes: Uint8Array): string {
  return Array.from(bytes)
    .map((b) => b.toString(16).padStart(2, "0"))
    .join("");
}

/** Convert hex string to Uint8Array */
export function fromHex(hex: string): Uint8Array {
  const bytes = new Uint8Array(hex.length / 2);
  for (let i = 0; i < hex.length; i += 2) {
    bytes[i / 2] = parseInt(hex.substring(i, i + 2), 16);
  }
  return bytes;
}

/** Sign a payload object with Ed25519 secret key. Returns hex signature. */
export function signPayload(payload: unknown, secretKey: Uint8Array): string {
  const msg = new TextEncoder().encode(canonicalJson(payload));
  const sig = nacl.sign.detached(msg, secretKey);
  return toHex(sig);
}

/** Verify a hex signature against a payload and hex public key. */
export function verifySignature(
  payload: unknown,
  signatureHex: string,
  publicKeyHex: string
): boolean {
  try {
    const msg = new TextEncoder().encode(canonicalJson(payload));
    const sig = fromHex(signatureHex);
    const pk = fromHex(publicKeyHex);
    return nacl.sign.detached.verify(msg, sig, pk);
  } catch {
    return false;
  }
}

/** Generate a new Ed25519 keypair. Returns { publicKey, secretKey } as Uint8Arrays. */
export function generateKeypair() {
  return nacl.sign.keyPair();
}

File: apps/api/app/crypto.py
"""Ed25519 signing and verification using PyNaCl (same algorithm as tweetnacl)."""

from __future__ import annotations

import nacl.signing
import nacl.encoding
from typing import Any

from .canonical_json import canonical_json


def to_hex(data: bytes) -> str:
    return data.hex()


def from_hex(hex_str: str) -> bytes:
    return bytes.fromhex(hex_str)


def sign_payload(payload: Any, secret_key: bytes) -> str:
    """Sign a payload object with Ed25519 secret key. Returns hex signature."""
    msg = canonical_json(payload).encode("utf-8")
    signing_key = nacl.signing.SigningKey(secret_key[:32])
    signed = signing_key.sign(msg)
    return to_hex(signed.signature)


def verify_signature(payload: Any, signature_hex: str, public_key_hex: str) -> bool:
    """Verify a hex signature against a payload and hex public key."""
    try:
        msg = canonical_json(payload).encode("utf-8")
        sig = from_hex(signature_hex)
        pk = from_hex(public_key_hex)
        verify_key = nacl.signing.VerifyKey(pk)
        verify_key.verify(msg, sig)
        return True
    except Exception:
        return False


def generate_keypair() -> tuple[bytes, bytes]:
    """Generate Ed25519 keypair. Returns (public_key, secret_key) as raw bytes."""
    signing_key = nacl.signing.SigningKey.generate()
    return bytes(signing_key.verify_key), bytes(signing_key) + bytes(signing_key.verify_key)


def keypair_from_seed(seed: bytes) -> tuple[bytes, bytes]:
    """Generate deterministic Ed25519 keypair from 32-byte seed.
    Returns (public_key_bytes, secret_key_bytes_64).
    The 64-byte secret key = seed + public_key (same as tweetnacl).
    """
    signing_key = nacl.signing.SigningKey(seed)
    pk = bytes(signing_key.verify_key)
    sk = bytes(signing_key) + pk
    return pk, sk
